core/src/main/scala/kafka/admin/TopicCommand.scala

```
@@ -17,248 +17,403 @@
       } else {
-        println(topic)
+        adminClient.listTopics(new ListTopicsOptions().listInternal(true))
       }
+      doGetTopics(allTopics.names().get().asScala.toSeq.sorted, topicWhitelist, excludeInternalTopics)
     }
+
+    override def close(): Unit = adminClient.close()
   }
 
-  def deleteTopic(zkClient: KafkaZkClient, opts: TopicCommandOptions) {
-    val topics = getTopics(zkClient, opts)
-    val ifExists = opts.options.has(opts.ifExistsOpt)
-    if (topics.isEmpty && !ifExists) {
-      throw new IllegalArgumentException("Topic %s does not exist on ZK path %s".format(opts.options.valueOf(opts.topicOpt),
-          opts.options.valueOf(opts.zkConnectOpt)))
+  class ZookeeperTopicService(zkClient: KafkaZkClient) extends TopicService {
+
+    def this(zkConnect: Option[String]) {
+      this(KafkaZkClient(zkConnect.get, JaasUtils.isZkSecurityEnabled, 30000, 30000,
+        Int.MaxValue, Time.SYSTEM))
     }
-    topics.foreach { topic =>
+
+    override def createTopic(topic: CommandTopicPartition): Unit = {
+      val adminZkClient = new AdminZkClient(zkClient)
       try {
-        if (Topic.isInternal(topic)) {
-          throw new AdminOperationException("Topic %s is a kafka internal topic and is not allowed to be marked for deletion.".format(topic))
-        } else {
-          zkClient.createDeleteTopicPath(topic)
-          println("Topic %s is marked for deletion.".format(topic))
-          println("Note: This will have no impact if delete.topic.enable is not set to true.")
+        if (topic.hasReplicaAssignment)
+          adminZkClient.createTopicWithAssignment(topic.name, topic.configsToAdd, topic.replicaAssignment.get)
+        else
+          adminZkClient.createTopic(topic.name, topic.partitions.get, topic.replicationFactor, topic.configsToAdd, topic.rackAwareMode)
+        println("Created topic \"%s\".".format(topic.name))
+      } catch  {
+        case e: TopicExistsException => if (!topic.ifTopicDoesntExist()) throw e
```

What's the purpose of this if?

---

That is a weird logic to handle a conditional operation. :)
So there is an `--if-not-exists` option in TopicCommand which would create the topic if it doesn't exist. Now the implementation will try to create the topic anyways but specifying this flag will suppress the exception. I didn't refactor this too much because it'll be deprecated and didn't want to change its behavior. In `AdminClientTopicService.createTopic` it is implemented properly.

---

Also looking at the bigger picture I'm not 100% convinced that we'll need the mentioned flag because for more advanced logic people should use the AdminClient directly, so we might even deprecate it.